dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DockerRegistryAPIHelper.java

```diff
@@ -0,0 +1,236 @@
+ */
+
+package io.dockstore.webservice.helpers;
+
+import static io.dockstore.webservice.Constants.JWT_SECURITY_DEFINITION_NAME;
+
+import com.google.common.hash.Hashing;
+import com.google.gson.Gson;
+import io.dockstore.webservice.CacheHitListener;
+import io.dockstore.webservice.DockstoreWebserviceApplication;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.ws.rs.core.HttpHeaders;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DockerRegistryAPIHelper {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DockerRegistryAPIHelper.class);
+    private static final Gson GSON = new Gson();
+    private final OkHttpClient client;
+
+    public static final String DOCKER_V2_IMAGE_MANIFEST_MEDIA_TYPE = "application/vnd.docker.distribution.manifest.v2+json";
+    public static final String DOCKER_V2_IMAGE_MANIFEST_LIST_MEDIA_TYPE = "application/vnd.docker.distribution.manifest.list.v2+json";
+    public static final String OCI_IMAGE_MANIFEST_MEDIA_TYPE = "application/vnd.oci.image.manifest.v1+json";
+    public static final String OCI_IMAGE_INDEX_MEDIA_TYPE = "application/vnd.oci.image.index.v1+json";
+
+    public DockerRegistryAPIHelper() {
+        // this code is duplicate from DockstoreWebserviceApplication, except this is a lot faster for unknown reasons ...
+        OkHttpClient.Builder builder = new OkHttpClient().newBuilder();
+        builder.eventListener(new CacheHitListener(DockerRegistryAPIHelper.class.getSimpleName(), "Docker Registry HTTP API V2"));
+
+        // use general cache
+        builder.cache(DockstoreWebserviceApplication.getCache(null));
+        this.client = builder.build();
+    }
+
+    /**
+     * Get an anonymous token with pull access to make Docker Registry HTTP API V2 calls.
+     * Source for token request specs: https://docs.docker.com/registry/spec/auth/token/#requesting-a-token
+     *
+     * @param registryDockerPath
+     * @param repo
+     * @return token
+     */
+    public Optional<String> getDockerToken(String registryDockerPath, String repo) {
+        String getTokenURL = String.format("https://%s/token?scope=repository:%s:pull&service=%s", registryDockerPath, repo, registryDockerPath);
+        Request request = new Request.Builder().url(getTokenURL).build();
+
+        Response tokenResponse;
+        try {
+            tokenResponse = client.newCall(request).execute();
+        } catch (IOException ex) {
+            LOG.error("Could not send token request GET {}", getTokenURL, ex);
+            return Optional.empty();
+        }
+
+        if (tokenResponse.isSuccessful()) {
+            Map<String, String> tokenMap = GSON.fromJson(tokenResponse.body().charStream(), Map.class);
+            return Optional.of(tokenMap.get("token"));
```

Comment:
Is `token` always guaranteed to be present? Probably, but if not, use `Optional.ofNullable`.

Reply:
I think it should be present if it's a 200 response, but I used `Optional.ofNullable` anyway since it doesn't hurt